server/src/test/java/org/apache/kafka/server/EligibleLeaderReplicasIntegrationTest.java

```diff
@@ -0,0 +1,384 @@
+                ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, clusterInstance.bootstrapServers(),
+                ConsumerConfig.GROUP_ID_CONFIG, "test",
+                ConsumerConfig.FETCH_MAX_WAIT_MS_CONFIG, "10",
+                ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest",
+                ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName(),
+                ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName()))) {
+            String testTopicName = String.format("%s-%s", "testHighWatermarkShouldNotAdvanceIfUnderMinIsr", "ELR-test");
+            admin.updateFeatures(
+                Map.of(EligibleLeaderReplicasVersion.FEATURE_NAME,
+                    new FeatureUpdate(EligibleLeaderReplicasVersion.ELRV_1.featureLevel(), FeatureUpdate.UpgradeType.UPGRADE)),
+                new UpdateFeaturesOptions()).all().get();
+
+            admin.createTopics(List.of(new NewTopic(testTopicName, 1, (short) 4))).all().get();
+            clusterInstance.waitTopicCreation(testTopicName, 1);
+
+            ConfigResource configResource = new ConfigResource(ConfigResource.Type.TOPIC, testTopicName);
+            Collection<AlterConfigOp> ops = new ArrayList<>();
+            ops.add(new AlterConfigOp(new ConfigEntry(TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG, "3"), AlterConfigOp.OpType.SET));
+            Map<ConfigResource, Collection<AlterConfigOp>> configOps = Map.of(configResource, ops);
+            // alter configs on target cluster
+            admin.incrementalAlterConfigs(configOps).all().get();
+
+            TopicDescription testTopicDescription = admin.describeTopics(List.of(testTopicName))
+                .allTopicNames().get().get(testTopicName);
+            TopicPartitionInfo topicPartitionInfo = testTopicDescription.partitions().get(0);
+            List<Node> initialReplicas = topicPartitionInfo.replicas();
+            assertEquals(4, topicPartitionInfo.isr().size());
+            assertEquals(0, topicPartitionInfo.elr().size());
+            assertEquals(0, topicPartitionInfo.lastKnownElr().size());
+
+            consumer.subscribe(Set.of(testTopicName));
+            producer.send(new ProducerRecord<>(testTopicName, "0", "0")).get();
+            waitUntilOneMessageIsConsumed(consumer);
+
+            clusterInstance.shutdownBroker(initialReplicas.get(0).id());
+            clusterInstance.shutdownBroker(initialReplicas.get(1).id());
+
+            waitForIsrAndElr((isrSize, elrSize) -> isrSize == 2 && elrSize == 1, admin, testTopicName);
+
+            // Now the partition is under min ISR. HWM should not advance.
+            producer.send(new ProducerRecord<>(testTopicName, "1", "1")).get();
+            // We use a short sleep here to give the broker time to process the ISR change.
+            // To ensure the consumer sees the correct HWM state to avoid flakiness.
+            TimeUnit.MILLISECONDS.sleep(100);
```

Comment:
Perhaps we should use `Admin#listOffsets` to ensure the offset advances. That is more reliable than using `sleep`